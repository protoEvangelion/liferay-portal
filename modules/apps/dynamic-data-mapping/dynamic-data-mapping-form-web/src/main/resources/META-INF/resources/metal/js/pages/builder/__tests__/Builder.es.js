import Builder from '../Builder.es';
import {dom as MetalTestUtil} from 'metal-dom';

let component;
const spritemap = 'icons.svg';
let addButton;

describe(
	'Builder',
	() => {
		beforeEach(
			() => {
				jest.useFakeTimers();

				MetalTestUtil.enterDocument('<button id="addFieldButton"></button>');

				MetalTestUtil.enterDocument('<div class="ddm-translation-manager"></div>');

				MetalTestUtil.enterDocument('<div class="ddm-form-basic-info"></div>');

				addButton = document.querySelector('#addFieldButton');

				component = new Builder(
					{
						spritemap
					}
				);
			}
		);

		afterEach(
			() => {
				MetalTestUtil.exitDocument(addButton);
				if (component) {
					component.dispose();
				}
			}
		);

		it(
			'should render the default markup',
			() => {
				expect(component).toMatchSnapshot();
			}
		);

		it(
			'should continue to propagate the fieldAdded event',
			() => {
				const spy = jest.spyOn(component, 'emit');
				const {sidebar} = component.refs;
				const mockEvent = jest.fn();

				sidebar.emit('fieldAdded', mockEvent);

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith('fieldAdded', expect.anything());
			}
		);

		it(
			'should continue to propagate the fieldEdited event',
			() => {
				const spy = jest.spyOn(component, 'emit');
				const {sidebar} = component.refs;
				const mockEvent = jest.fn();

				sidebar.emit('fieldEdited', mockEvent);

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith('fieldEdited', expect.anything());
			}
		);

		it(
			'should continue to propagate the fieldMoved event',
			() => {
				const spy = jest.spyOn(component, 'emit');
				const {FormRenderer} = component.refs;
				const mockEvent = jest.fn();

				FormRenderer.emit('fieldMoved', mockEvent);

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith('fieldMoved', expect.anything());
			}
		);

		it(
			'should continue to propagate the deleteField event',
			() => {
				const spy = jest.spyOn(component, 'emit');
				const {FormRenderer} = component.refs;
				const mockEvent = jest.fn();

				FormRenderer.emit('deleteButtonClicked', mockEvent);

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith('deleteField', expect.anything());
			}
		);

		it(
			'should continue to propagate pagesUpdated event',
			() => {
				const spy = jest.spyOn(component, 'emit');
				const {FormRenderer} = component.refs;
				const mockEvent = jest.fn();

				FormRenderer.emit('pagesUpdated', mockEvent);

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith('pagesUpdated', expect.anything());
			}
		);

		it(
			'should continue to propagate pageAdded event',
			() => {
				const spy = jest.spyOn(component, 'emit');
				const {FormRenderer} = component.refs;
				const mockEvent = jest.fn();

				FormRenderer.emit('pageAdded', mockEvent);

				jest.runAllTimers();

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith('pagesUpdated', expect.anything());
			}
		);

		it(
			'should continue to propagate the fieldClicked event and open the sidebar',
			() => {
				const spy = jest.spyOn(component, 'emit');
				const {FormRenderer, sidebar} = component.refs;
				const mockEvent = jest.fn();

				FormRenderer.emit('fieldClicked', mockEvent);

				jest.runAllTimers();

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith('fieldClicked', expect.anything());
				expect(sidebar.state.show).toBeTruthy();
			}
		);

		it(
			'should continue to propagate the duplicateField event',
			() => {
				const spy = jest.spyOn(component, 'emit');
				const {FormRenderer} = component.refs;
				const mockEvent = jest.fn();

				FormRenderer.emit('duplicateButtonClicked', mockEvent);

				jest.runAllTimers();

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith('duplicateField', expect.anything());
			}
		);

		it(
			'should open sidebar when active page is changed and mode is "add"',
			() => {
				const {FormRenderer} = component.refs;
				const mode = 'add';

				FormRenderer.emit('activePageUpdated', {mode});

				jest.runAllTimers();

				expect(component).toMatchSnapshot();
			}
		);

		it(
			'should not open sidebar when the active page is changed and mode is not "add"',
			() => {
				const {FormRenderer} = component.refs;
				const mockEvent = jest.fn();

				FormRenderer.emit('activePageUpdated', mockEvent);

				jest.runAllTimers();

				expect(component).toMatchSnapshot();
			}
		);
	}
);