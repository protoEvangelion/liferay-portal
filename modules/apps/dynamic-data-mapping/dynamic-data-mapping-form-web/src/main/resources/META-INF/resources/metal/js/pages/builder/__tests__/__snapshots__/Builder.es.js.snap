// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder should not open sidebar when the active page is changed and mode is not "add" 1`] = `
<div>
  <div class="container">
    <div class="sheet">
      <div class="position-relative">
        <div class="ddm-form-header">
          <div class="ddm-page-settings" role="group">
            <div class="dropdown dropdown-action">
              <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn component-action btn-unstyled" title="actions" type="button">
                <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v" focusable="false">
                  <title>ellipsis-v</title>
                  <use xlink:href="icons.svg#ellipsis-v"></use>
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="sidebar-container" ref="sidebar">
    <div class="sidebar sidebar-light">
      <nav class="component-tbar tbar">
        <div class="container-fluid">
          <ul class="tbar-nav">
            <li class="tbar-item">
              <a class="component-action" href="#1" ref="close" role="button">
                <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                  <use xlinkHref="icons.svg#times"></use>
                </svg>
              </a>
            </li>
          </ul>
        </div>
      </nav>
      <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
        <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
          <span class="navbar-text-truncate">Details</span>
          <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
            <use xlinkHref="icons.svg#caret-bottom"></use>
          </svg>
        </a>
        <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
          <ul class="nav navbar-nav" role="tablist">
            <li class="nav-item" data-index="0" ref="tab0">
              <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
                <span class="navbar-text-truncate">Elements</span>
              </a>
            </li>
          </ul>
        </div>
      </nav>
      <div class="ddm-sidebar-body"></div>
    </div>
  </div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <button class="btn dropdown-item btn-sm btn-link" aria-label="add-new-page" type="button">add-new-page</button>
    </li>
    <li>
      <button class="btn dropdown-item btn-sm btn-link" aria-label="reset-page" type="button">reset-page</button>
    </li>
  </ul>
</div>
`;

exports[`Builder should open sidebar when active page is changed and mode is "add" 1`] = `
<div>
  <div class="container">
    <div class="sheet">
      <div class="position-relative">
        <div class="ddm-form-header">
          <div class="ddm-page-settings" role="group">
            <div class="dropdown dropdown-action">
              <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn component-action btn-unstyled" title="actions" type="button">
                <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v" focusable="false">
                  <title>ellipsis-v</title>
                  <use xlink:href="icons.svg#ellipsis-v"></use>
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="sidebar-container show" ref="sidebar">
    <div class="sidebar sidebar-light">
      <nav class="component-tbar tbar">
        <div class="container-fluid">
          <ul class="tbar-nav">
            <li class="tbar-item tbar-item-expand text-left">
              <div class="tbar-section">
                <span class="text-truncate-inline">
                  <span class="text-truncate">Add Elements</span>
                </span>
              </div>
            </li>
            <li class="tbar-item">
              <a class="component-action" href="#1" ref="close" role="button">
                <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                  <use xlinkHref="icons.svg#times"></use>
                </svg>
              </a>
            </li>
          </ul>
        </div>
      </nav>
      <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
        <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
          <span class="navbar-text-truncate">Details</span>
          <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
            <use xlinkHref="icons.svg#caret-bottom"></use>
          </svg>
        </a>
        <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
          <ul class="nav navbar-nav" role="tablist">
            <li class="nav-item" data-index="0" ref="tab0">
              <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
                <span class="navbar-text-truncate">Elements</span>
              </a>
            </li>
          </ul>
        </div>
      </nav>
      <div class="ddm-sidebar-body"></div>
    </div>
  </div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <button class="btn dropdown-item btn-sm btn-link" aria-label="add-new-page" type="button">add-new-page</button>
    </li>
    <li>
      <button class="btn dropdown-item btn-sm btn-link" aria-label="reset-page" type="button">reset-page</button>
    </li>
  </ul>
</div>
`;

exports[`Builder should render the default markup 1`] = `
<div>
  <div class="container">
    <div class="sheet">
      <div class="position-relative">
        <div class="ddm-form-header">
          <div class="ddm-page-settings" role="group">
            <div class="dropdown dropdown-action">
              <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn component-action btn-unstyled" title="actions" type="button">
                <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v" focusable="false">
                  <title>ellipsis-v</title>
                  <use xlink:href="icons.svg#ellipsis-v"></use>
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="sidebar-container" ref="sidebar">
    <div class="sidebar sidebar-light">
      <nav class="component-tbar tbar">
        <div class="container-fluid">
          <ul class="tbar-nav">
            <li class="tbar-item">
              <a class="component-action" href="#1" ref="close" role="button">
                <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                  <use xlinkHref="icons.svg#times"></use>
                </svg>
              </a>
            </li>
          </ul>
        </div>
      </nav>
      <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
        <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
          <span class="navbar-text-truncate">Details</span>
          <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
            <use xlinkHref="icons.svg#caret-bottom"></use>
          </svg>
        </a>
        <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
          <ul class="nav navbar-nav" role="tablist">
            <li class="nav-item" data-index="0" ref="tab0">
              <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
                <span class="navbar-text-truncate">Elements</span>
              </a>
            </li>
          </ul>
        </div>
      </nav>
      <div class="ddm-sidebar-body"></div>
    </div>
  </div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <button class="btn dropdown-item btn-sm btn-link" aria-label="add-new-page" type="button">add-new-page</button>
    </li>
    <li>
      <button class="btn dropdown-item btn-sm btn-link" aria-label="reset-page" type="button">reset-page</button>
    </li>
  </ul>
</div>
`;
